pkg/repair/repaircontroller.go

```diff
@@ -0,0 +1,119 @@
+package repair
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"istio.io/pkg/log"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	client "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	"k8s.io/client-go/util/workqueue"
+)
+
+type RepairController struct {
+	clientset     client.Interface
+	workQueue     workqueue.RateLimitingInterface
+	podController cache.Controller
+
+	reconciler BrokenPodReconciler
+}
+
+func NewRepairController(reconciler BrokenPodReconciler) (*RepairController, error) {
+	c := &RepairController{
+		clientset:  reconciler.client,
+		workQueue:  workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		reconciler: reconciler,
+	}
+
+	podListWatch := cache.NewFilteredListWatchFromClient(
+		c.clientset.CoreV1().RESTClient(),
+		"pods",
+		metav1.NamespaceAll,
+		func(options *metav1.ListOptions) {
+			labelSelectors := []string{}
+			fieldSelectors := []string{}
+
+			for _, ls := range []string{options.LabelSelector, reconciler.Filters.LabelSelectors} {
+				if ls != "" {
+					labelSelectors = append(labelSelectors, ls)
+				}
+			}
+			for _, fs := range []string{options.FieldSelector, reconciler.Filters.LabelSelectors} {
+				if fs != "" {
+					fieldSelectors = append(fieldSelectors, fs)
+				}
+			}
+			options.LabelSelector = strings.Join(labelSelectors, ",")
+			options.FieldSelector = strings.Join(fieldSelectors, ",")
+		},
+	)
+
+	_, c.podController = cache.NewInformer(podListWatch, &v1.Pod{}, 0, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(newObj interface{}) {
+			c.workQueue.AddRateLimited(newObj)
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			c.workQueue.AddRateLimited(newObj)
+		},
+	})
+
+	return c, nil
+}
+
+func (rc *RepairController) Run(stopCh <-chan struct{}) {
+	go rc.podController.Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, rc.podController.HasSynced) {
+		runtime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return
+	}
+
+	go wait.Until(
+		func() {
+			for rc.processNextItem() {
```

Comment:
what is the point of return code from processNextItem? seems like we always just loop forever regardless

Reply:
I'm not sure what you are asking -- if `rc.processNextItem()` returns `false`, the loop exits.

Reply:
wait.Until will restart it though? you basically have `for{ for{ } }` is my understanding

Reply:
Oh, OK. This was based on https://github.com/kubernetes/client-go/blob/master/examples/workqueue/main.go, and that is how it works.

I'll remove the for loop.

Reply:
Actually, it doesn't appear to work, since wait.Until delays each run of the func by 1 second.